Renderer/CKEditorRenderer.php

```diff
@@ -25,16 +24,44 @@
 class CKEditorRenderer implements CKEditorRendererInterface
 {
     /**
-     * @var ContainerInterface
+     * @var JsonBuilder
      */
-    private $container;
+    private $jsonBuilder;
 
     /**
-     * @param ContainerInterface $container
+     * @var RouterInterface
      */
-    public function __construct(ContainerInterface $container)
+    private $router;
+
+    /**
+     * @var Packages
+     */
+    private $assetsPackages;
+
+    /**
+     * @var EngineInterface
+     */
+    private $templating;
+
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @param JsonBuilder     $jsonBuilder
+     * @param RouterInterface $router
+     * @param Packages        $packages
+     * @param RequestStack    $requestStack
+     * @param EngineInterface $templating
+     */
+    public function __construct(JsonBuilder $jsonBuilder, RouterInterface $router, Packages $packages, RequestStack $requestStack, EngineInterface $templating)
```

Comment:
Changing the constructor signature is a breaking change. What can we do here to avoid this?

Reply:
You should be using the `CKEditorRendererInterface`, not extending the base class. You would have to make all the services used by the render public otherwise.

Reply:
I know, but the class is not `final` so it looks to be a valid (but bad) extension point

Reply:
So what to do? Injecting the services is the cleanest way to avoid `$container->get()` calls

Reply:
I would allow to use a ServiceLocator alongside the old ContainerInterface. If a container is provided and it contains the ServiceLocator then fetch it from the container. If there is no service locator keep going using ContainerInterface::get. Then in the DI, we have to inject the ServiceLocator if using a Symfony version supporting it or inject the container (< 3.3).

Reply:
@ElectricMaxww I can send a PR to your branch if you want.

Reply:
that is not the point. The point is a kind of a philosophy. I personally do not like `ServiceLocators`, cause i loose type hints. I am back in a `$bla->get('blubb')`. Sure, we would be near at the former solution. But i think #342 goes that way. So lets decide which approach to use. I prefer DI over `ServiceLocators`, but i can be wrong.

Reply:
I prefer too but that’s for BC purpose. After that PR, we will be able to deprecate the ContainerInterface and ServiceLocator to use plain DI. But not in this PR, to keep it as small as possible.

Reply:
This would mean to re-work tests again, as the dependencies are served through the service locator. 

if you have time, you can do a PR against mine, otherwise i would do tonight.

Reply:
@fbourigault the easiest way to fix this is to simply make all the services public and then change it back in the new release. Considering you need the `locale` parameter anyway, it would be best to leave the container as it is. That or we can just roll these changes into a new release and leave `6.x` incompatibile with SF4.

Reply:
IMHO we should allow the old signature with this new one. No need of `ServiceLocator`.
@ElectricMaxxx I could make a separate PR that deprecate the use of `ContainerInterface` as first argument and use the signature you introduced here. This would make the whole thing easier to review.

Reply:
I ended by doing a PR to your branch. https://github.com/ElectricMaxxx/IvoryCKEditorBundle/pull/1

I didn't used a `ServiceLocator` as in Symfony 3.4, the user will get deprecation notices from Symfony about getting services from the container.